### ISSUE TITLE
Should be Train_steps_per_epoch changed according to training set size and batch size?

### ISSUE THREAD

**Issue by javierpastorfernandez** [stardist/stardist#160]
Good evening, 

I was revisiting some of the trainings I performed during my work this semester and I come across one technical doubt.
StarDist allows you to specify batch size, number of epochs, and steps_per_epoch, as shown below.

```python
Config2D(axes='YXC', backbone='unet', grid=(2, 2), n_channel_in=3, n_channel_out=33, n_classes=None, n_dim=2, n_rays=32, net_conv_after_unet=128, net_input_shape=(None, None, 3), net_mask_shape=(None, None, 1), train_background_reg=0.0001, train_batch_size=4, train_checkpoint='weights_best.h5', train_checkpoint_epoch='weights_now.h5', train_checkpoint_last='weights_last.h5', train_class_weights=(1, 1), train_completion_crop=32, train_dist_loss='mae', train_epochs=400, train_foreground_only=0.9, train_learning_rate=0.0003, train_loss_weights=(1, 0.2), train_n_val_patches=None, train_patch_size=(512, 512), train_reduce_lr={'factor': 0.5, 'patience': 40, 'min_delta': 0}, train_sample_cache=True, train_shape_completion=True, train_steps_per_epoch=100, train_tensorboard=True, unet_activation='relu', unet_batch_norm=False, unet_dropout=0.0, unet_kernel_size=(3, 3), unet_last_activation='relu', unet_n_conv_per_depth=2, unet_n_depth=3, unet_n_filter_base=32, unet_pool=(2, 2), unet_prefix='', use_gpu=False)
{'axes': 'YXC',
 'backbone': 'unet',
 'grid': (2, 2),
 'n_channel_in': 3,
 'n_channel_out': 33,
 'n_classes': None,
 'n_dim': 2,
 'n_rays': 32,
 'net_conv_after_unet': 128,
 'net_input_shape': (None, None, 3),
 'net_mask_shape': (None, None, 1),
 'train_background_reg': 0.0001,
 'train_batch_size': 4,
 'train_checkpoint': 'weights_best.h5',
 'train_checkpoint_epoch': 'weights_now.h5',
 'train_checkpoint_last': 'weights_last.h5',
 'train_class_weights': (1, 1),
 'train_completion_crop': 32,
 'train_dist_loss': 'mae',
 'train_epochs': 400,
 'train_foreground_only': 0.9,
 'train_learning_rate': 0.0003,
 'train_loss_weights': (1, 0.2),
 'train_n_val_patches': None,
 'train_patch_size': (512, 512),
 'train_reduce_lr': {'factor': 0.5, 'min_delta': 0, 'patience': 40},
 'train_sample_cache': True,
 'train_shape_completion': True,
 'train_steps_per_epoch': 100,
 'train_tensorboard': True,
 'unet_activation': 'relu',
 'unet_batch_norm': False,
 'unet_dropout': 0.0,
 'unet_kernel_size': (3, 3),
 'unet_last_activation': 'relu',
 'unet_n_conv_per_depth': 2,
 'unet_n_depth': 3,
 'unet_n_filter_base': 32,
 'unet_pool': (2, 2),
 'unet_prefix': '',
 'use_gpu': False}
```

Nevertheless, in "model2d.py", when it calls the Keras fit function it seems only to specify the epochs and iterations/steps_per_epoch. Batch size parameter seems to be used during the Data Generation Functions.

```python
history = fit(iter(data_train), validation_data=data_val,
                      epochs=epochs, steps_per_epoch=steps_per_epoch,
                      callbacks=self.callbacks, verbose=1)
        self._training_finished()
```

I read online some people set `steps_per_epoch= training set size / batch size`,  as one epoch is defined whenever the model is trained along the whole dataset. 

I am worried I did something wrong. For all my trainings I kept steps_per_epoch parameter unchanged to its default value=100. 

Thank you for your help!
Javier

**Comment by uschmidt83** [comment]
> I read online some people set `steps_per_epoch= training set size / batch size`, as one epoch is defined whenever the model is trained along the whole dataset.

Yes, that's a common definition.

> I am worried I did something wrong. For all my trainings I kept steps_per_epoch parameter unchanged to its default value=100.

No, you didn't. We don't recommend to change the `steps_per_epoch` parameter as you mentioned above. Some people have really small datasets, i.e. would only do a few steps per epoch. However, the `train_reduce_lr` parameter (with `patience ` given in epochs) will cause the learning rate to be automatically adjusted when training plateaus. Hence, these parameters are intertwined and we chose default parameters that would lead to well trained models regardless of dataset size.

Best,
Uwe

**Comment by uschmidt83** [comment]
In the future, please ask questions like this in the [image.sc forum](https://forum.image.sc/tag/stardist) and use the tag `stardist`.

**Comment by javierpastorfernandez** [comment]
Thank you very much!!!
